kiwipy/rmq/communicator.py

```diff
@@ -334,24 +347,79 @@ async def __aexit__(self, exc_type, exc_val, exc_tb):
         await self.disconnect()
 
     def __str__(self):
-        return "RMQCommunicator ({})".format(self._connection)
+        return "RMQCommunicator({})".format(self._connection)
 
     @property
     def loop(self):
         return self._connection.loop
 
+    def get_default_task_queue(self) -> Optional[tasks.RmqTaskQueue]:
+        if not self.connected():
+            return None
+
+        if self._default_task_queue is None:
+            self._default_task_queue = tasks.RmqTaskQueue(self._connection,
```

Comment:
Should these be connected explicitly when instantiated, like is done in the `connect` method or is this done lazily somehow?

If I understand the logic correctly, the idea is that you always first have to call `connect` which will call these methods, which will construct the task queue, sub and pub, and then calls connect on them. Is there a possibility that if `connect` successfully opens the connection, but than fails or gets interrupted at line 408. Calling `get_message_subscriber` will then happily return the subscriber, but it won't actually be connected. Should this method just lazily do the connection? Something like this:
```python
    def get_default_task_queue(self) -> Optional[tasks.RmqTaskQueue]:
        if not self.connected():
            return None

        if self._default_task_queue is None:
            self._default_task_queue = tasks.RmqTaskQueue(self._connection,
                                                         exchange_name=self._task_exchange,
                                                          queue_name=self._task_queue,
                                                          decoder=self._decoder,
                                                          encoder=self._encoder,
                                                          prefetch_size=self._task_prefetch_size,
                                                          prefetch_count=self._task_prefetch_count,
                                                          testing_mode=self._testing_mode)
        if self._default_task_queue.is_closed:
            self._default_task_queue.connect()

         return self._default_task_queue
```

Reply:
I'm not sure I quiet understand the flow of the connection failure you outlined but let me sum up my thinking on this and maybe it'll become clearer.

Basically, for the communicator classes the user has to explicitly instantiate _and_ call the `connect` member function but the preferred way to do this is, of course, the free (factory) function which does both for you.  I don't think the communicator should be lazily connecting if the user forgot because this is a resource expensive and error prone operation and should be done explicitly.

Is there a flow that you think we should put more error checking around?

Reply:
This will be addressed by PR #53